lib/AST/ASTContext.cpp

```diff
@@ -4187,8 +4187,12 @@ CanSILBoxType SILBoxType::get(ASTContext &C,
                               SILLayout *Layout,
                               SubstitutionList Args) {
   llvm::FoldingSetNodeID id;
+
+  // Canonicalize substitutions.
+  Args = getCanonicalSubstitutionList(Args);
```

Comment:
It looks like here we allocate memory for the substitutions inside the SILBoxType also, so an out parameter vector would be better than having getCanonicalSubstitutionList() allocate from context.